devices/src/virtio/vsock/packet.rs

```diff
@@ -342,3 +342,292 @@ impl VsockPacket {
+                .flags
+                .set(VIRTQ_DESC_F_WRITE);
+            expect_asm_error!(tx, test_ctx, handler_ctx, VsockError::UnreadableDescriptor);
+        }
+
+        // Test case: header descriptor has inssuficient space to hold the packet header.
+        {
+            create_context!(test_ctx, handler_ctx);
+            handler_ctx.guest_txvq.dtable[0]
+                .len
+                .set(VSOCK_PKT_HDR_SIZE as u32 - 1);
+            expect_asm_error!(tx, test_ctx, handler_ctx, VsockError::HdrDescTooSmall(_));
+        }
+
+        // Test case: zero-length TX packet.
+        {
+            create_context!(test_ctx, handler_ctx);
+            set_pkt_len(0, &handler_ctx.guest_txvq.dtable[0], &test_ctx.mem);
+            let mut pkt = VsockPacket::from_tx_virtq_head(
+                &handler_ctx.handler.txvq.pop(&test_ctx.mem).unwrap(),
+            )
+            .unwrap();
+            assert!(pkt.buf().is_none());
+            assert!(pkt.buf_mut().is_none());
+        }
+
+        // Test case: TX packet has more data than we can handle.
+        {
+            create_context!(test_ctx, handler_ctx);
+            set_pkt_len(
+                MAX_PKT_BUF_SIZE as u32 + 1,
+                &handler_ctx.guest_txvq.dtable[0],
+                &test_ctx.mem,
+            );
+            expect_asm_error!(tx, test_ctx, handler_ctx, VsockError::InvalidPktLen(_));
+        }
+
+        // Test case:
+        // - packet header advertises some data length; and
+        // - the data descriptor is missing.
+        {
+            create_context!(test_ctx, handler_ctx);
+            set_pkt_len(1024, &handler_ctx.guest_txvq.dtable[0], &test_ctx.mem);
+            handler_ctx.guest_txvq.dtable[0].flags.set(0);
+            expect_asm_error!(tx, test_ctx, handler_ctx, VsockError::BufDescMissing);
+        }
+
+        // Test case: error on write-only buf descriptor.
+        {
+            create_context!(test_ctx, handler_ctx);
+            handler_ctx.guest_txvq.dtable[1]
+                .flags
+                .set(VIRTQ_DESC_F_WRITE);
+            expect_asm_error!(tx, test_ctx, handler_ctx, VsockError::UnreadableDescriptor);
+        }
+
+        // Test case: the buffer descriptor cannot fit all the data advertised by the the
+        // packet header `len` field.
+        {
+            create_context!(test_ctx, handler_ctx);
+            set_pkt_len(8 * 1024, &handler_ctx.guest_txvq.dtable[0], &test_ctx.mem);
+            handler_ctx.guest_txvq.dtable[1].len.set(4 * 1024);
+            expect_asm_error!(tx, test_ctx, handler_ctx, VsockError::BufDescTooSmall);
+        }
+    }
+
+    #[test]
+    fn test_rx_packet_assembly() {
+        // Test case: successful RX packet assembly.
+        {
+            create_context!(test_ctx, handler_ctx);
+            let _pkt = VsockPacket::from_rx_virtq_head(
+                &handler_ctx.handler.rxvq.pop(&test_ctx.mem).unwrap(),
+            )
+            .unwrap();
```

Comment:
I think it makes sens to check that `from_rx_virtq_head` returns a packet that is comprised of parts (header and data) that have the right length i.e.:
```
assert_eq!(_pkt.hdr().len(), VSOCK_PKT_HDR_SIZE);
assert_eq!(
       _pkt.buf().unwrap().len(),
       handler_ctx.guest_rxvq.dtable[1].len.get() as usize
);
assert_eq!(
                pkt.buf_size,
                handler_ctx.guest_txvq.dtable[1].len.get() as usize
            );
```
(nit: eventually turn _pkt into pkt)